File: packages/audit-trail/examples/basic_logging.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * basic_logging.ts — Demonstrates core AuditLogger usage.
 *
 * Shows how to:
 * - Create a logger (defaults to in-memory storage)
 * - Log governance decisions (permitted and denied)
 * - Query the log with filters
 * - Count records
 *
 * Run: npx tsx examples/basic_logging.ts
 */

import { AuditLogger } from "../typescript/src/index.js";
import type { AuditFilter } from "../typescript/src/index.js";

async function main(): Promise<void> {
  const logger = new AuditLogger();

  console.log("=== AumOS Audit Trail — Basic Logging Example ===\n");

  // Log a series of governance decisions across two agents.
  const decisions = [
    {
      agentId: "agent-crm-001",
      action: "read_customer_record",
      permitted: true,
      trustLevel: 3,
      requiredLevel: 2,
      reason: "Trust level meets requirement for read access",
    },
    {
      agentId: "agent-crm-001",
      action: "export_customer_data",
      permitted: false,
      trustLevel: 3,
      requiredLevel: 5,
      reason: "Trust level insufficient for bulk data export",
    },
    {
      agentId: "agent-crm-001",
      action: "send_email",
      permitted: true,
      trustLevel: 3,
      requiredLevel: 3,
      budgetUsed: 0.02,
      budgetRemaining: 9.98,
      reason: "Action permitted within budget",
    },
    {
      agentId: "agent-billing-002",
      action: "read_invoice",
      permitted: true,
      trustLevel: 4,
      requiredLevel: 2,
      metadata: { invoiceId: "INV-2026-0042" },
    },
    {
      agentId: "agent-billing-002",
      action: "issue_refund",
      permitted: false,
      trustLevel: 4,
      requiredLevel: 6,
      reason: "Refund issuance requires maximum trust level",
      metadata: { amount: 150.0, currency: "USD" },
    },
  ];

  console.log("Logging decisions...");
  const records = [];
  for (const decision of decisions) {
    const record = await logger.log(decision);
    records.push(record);
    const status = record.permitted ? "PERMITTED" : "DENIED  ";
    console.log(`  [${status}] ${record.agentId} -> ${record.action} | hash: ${record.recordHash.slice(0, 16)}...`);
  }

  const totalCount = await logger.count();
  console.log(`\nTotal records: ${totalCount}`);

  // Query all denied decisions.
  console.log("\n--- Denied decisions ---");
  const deniedFilter: AuditFilter = { permitted: false };
  const denied = await logger.query(deniedFilter);
  for (const record of denied) {
    console.log(`  ${record.agentId} -> ${record.action}: ${record.reason ?? "no reason"}`);
  }

  // Query decisions for a specific agent.
  console.log("\n--- Decisions for agent-crm-001 ---");
  const agentFilter: AuditFilter = { agentId: "agent-crm-001" };
  const agentRecords = await logger.query(agentFilter);
  for (const record of agentRecords) {
    console.log(`  [${record.permitted ? "OK" : "NO"}] ${record.action}`);
  }

  // Show the first record's full structure.
  console.log("\n--- First record (full) ---");
  const firstRecord = records[0];
  if (firstRecord !== undefined) {
    console.log(JSON.stringify(firstRecord, null, 2));
  }
}

main().catch((error: unknown) => {
  console.error("Error:", error);
  process.exit(1);
});

File: packages/audit-trail/examples/basic_logging.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 MuVeraAI Corporation

"""
basic_logging.py — Demonstrates core AuditLogger usage.

Shows how to:
- Create a logger (defaults to in-memory storage)
- Log governance decisions (permitted and denied)
- Query the log with filters
- Count records

Run: python examples/basic_logging.py
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path

# Allow running directly from the examples directory.
sys.path.insert(0, str(Path(__file__).parent.parent / "python" / "src"))

from audit_trail import AuditLogger, AuditFilter, GovernanceDecisionInput


async def main() -> None:
    logger = AuditLogger()

    print("=== AumOS Audit Trail — Basic Logging Example ===\n")

    decisions = [
        GovernanceDecisionInput(
            agent_id="agent-crm-001",
            action="read_customer_record",
            permitted=True,
            trust_level=3,
            required_level=2,
            reason="Trust level meets requirement for read access",
        ),
        GovernanceDecisionInput(
            agent_id="agent-crm-001",
            action="export_customer_data",
            permitted=False,
            trust_level=3,
            required_level=5,
            reason="Trust level insufficient for bulk data export",
        ),
        GovernanceDecisionInput(
            agent_id="agent-crm-001",
            action="send_email",
            permitted=True,
            trust_level=3,
            required_level=3,
            budget_used=0.02,
            budget_remaining=9.98,
            reason="Action permitted within budget",
        ),
        GovernanceDecisionInput(
            agent_id="agent-billing-002",
            action="read_invoice",
            permitted=True,
            trust_level=4,
            required_level=2,
            metadata={"invoice_id": "INV-2026-0042"},
        ),
        GovernanceDecisionInput(
            agent_id="agent-billing-002",
            action="issue_refund",
            permitted=False,
            trust_level=4,
            required_level=6,
            reason="Refund issuance requires maximum trust level",
            metadata={"amount": 150.0, "currency": "USD"},
        ),
    ]

    print("Logging decisions...")
    records = []
    for decision in decisions:
        record = await logger.log(decision)
        records.append(record)
        status = "PERMITTED" if record.permitted else "DENIED   "
        print(
            f"  [{status}] {record.agent_id} -> {record.action}"
            f" | hash: {record.record_hash[:16]}..."
        )

    total_count = await logger.count()
    print(f"\nTotal records: {total_count}")

    # Query all denied decisions.
    print("\n--- Denied decisions ---")
    denied = await logger.query(AuditFilter(permitted=False))
    for record in denied:
        print(f"  {record.agent_id} -> {record.action}: {record.reason or 'no reason'}")

    # Query decisions for a specific agent.
    print("\n--- Decisions for agent-crm-001 ---")
    agent_records = await logger.query(AuditFilter(agent_id="agent-crm-001"))
    for record in agent_records:
        marker = "OK" if record.permitted else "NO"
        print(f"  [{marker}] {record.action}")

    # Show the first record's full structure.
    print("\n--- First record (full) ---")
    if records:
        import json
        print(json.dumps(records[0].model_dump(mode="json"), indent=2))


if __name__ == "__main__":
    asyncio.run(main())
